allennlp/interpret/attackers/input_reduction.py

```diff
@@ -0,0 +1,147 @@
+
+            # Set num_ignore_tokens for NER and build token mask
+            else:
+                num_ignore_tokens, tag_mask, original_tags = \
+                    get_ner_tags_and_mask(current_instances, input_field_to_attack, ignore_tokens)
+            current_tokens = current_instances[0][input_field_to_attack].tokens
+            smallest_idx = -1
+            # keep removing tokens
+            while len(current_instances[0][input_field_to_attack]) >= num_ignore_tokens:
+                # get gradients and predictions
+                grads, outputs = self.predictor.get_gradients(current_instances)
+                for output in outputs:
+                    if isinstance(outputs[output], torch.Tensor):
+                        outputs[output] = outputs[output].detach().cpu().numpy().squeeze().squeeze()
+                    elif isinstance(outputs[output], list):
+                        outputs[output] = outputs[output][0]
+
+                # Check if any fields have changed, if so, break loop
+                if "tags" not in current_instances[0]:
+                    equal = True
+                    for field in fields_to_check:
+                        if field in current_instances[0].fields:
+                            equal = equal and current_instances[0][field].__eq__(fields_to_check[field])
+                        else:
+                            equal = equal and outputs[field] == fields_to_check[field]
+                    if not equal:
+                        break
+
+                # special case for sentence tagging (we have tested NER)
+                else:
+                    if smallest_idx != -1:
+                        del tag_mask[smallest_idx]
+                    cur_tags = [outputs["tags"][x] for x in range(len(outputs["tags"])) if tag_mask[x]]
+                    if cur_tags != original_tags:
+                        break
+
+                # remove a token from the input
+                current_tokens = list(current_instances[0][input_field_to_attack].tokens)
+                current_instances, smallest_idx = \
+                    remove_one_token(grads[grad_input_field],
+                                     current_instances,
+                                     input_field_to_attack, ignore_tokens)
+
+            final_tokens.append(current_tokens)
+        return sanitize({"final": final_tokens, "original": original_tokens})
+
+def remove_one_token(grads: np.ndarray,
+                     instances: List[Instance] = None,
+                     input_field_to_attack: str = 'tokens',
+                     ignore_tokens: List[str] = ["@@NULL@@"]) -> Tuple[List[Instance], int]:
+    """
+    Finds the token with the smallest gradient and removes it.
+    """
+    # Compute L2 norm of all grads.
+    grads_mag = [np.sqrt(grad.dot(grad)) for grad in grads]
+
+    # Skip all ignore_tokens by setting grad to infinity
+    for tok_idx, tok in enumerate(instances[0][input_field_to_attack].tokens):
```

Comment:
mypy is upset because you are assuming that `instance[input_field_to_attack]` is a text field (and has a .tokens attribute). you need to either use a cast or do something like

```
field = instances[0][input_field_to_attack]
tokens = getattr(field, 'tokens')
for idx, token in enumerate(tokens): ...
```

Reply:
This looks like the same issue as [here](https://github.com/allenai/allennlp/pull/3032#discussion_r301815275), and I'd recommend the same solution.